File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int k) {
        if(head==NULL) return NULL;
        int cnt=0;
        ListNode* curr=head;
        while(curr!=NULL){
            cnt++;
            curr=curr->next;
        }

        if(cnt<k){
            return head;
        }


        if (cnt == k) {
            ListNode* toDelete = head;
            head = head->next;
            delete toDelete;
            return head;
        }


        int pos=cnt-k;
        curr=head;
        ListNode* prev=head;
        while(pos--){
            prev=curr;
            curr=curr->next;
        }

        prev->next=curr->next;
        curr->next=NULL;
        delete curr;

        return head;

    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        group_prev = dummy

        while True:
            # 1. Check that k nodes exist (read-only walk)
            kth = group_prev
            for _ in range(k):
                kth = kth.next
                if not kth:
                    return dummy.next

            next_group_start = kth.next
            group_tail = group_prev.next   # first node now, last node after reversal

            # 2. Reverse exactly k nodes
            prev, curr = next_group_start, group_tail
            while curr != next_group_start:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt

            # 3. Reconnect and advance
            group_prev.next = kth          # kth is the group's new head
            group_prev = group_tail
        
